tests/test_ptmalloc.py

```diff
@@ -0,0 +1,282 @@
+    sc = s.copy()
+    p1 = s.heap.malloc(0x1000)
+    nose.tools.assert_equals(p1, 0)
+    nose.tools.assert_true(same_heap_states(s, sc))
+
+def test_needs_space_for_metadata():
+    for arch in ('X86', 'AMD64'):
+        yield run_needs_space_for_metadata, arch
+
+
+def run_unusable_amount_returns_null(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.heap.malloc(0x1000 - 4 * s.heap._chunk_size_t_size)
+    sc = s.copy()
+    p = s.heap.malloc(1)
+    nose.tools.assert_equals(p, 0)
+    nose.tools.assert_true(same_heap_states(s, sc))
+
+def test_unusable_amount_returns_null():
+    for arch in ('X86', 'AMD64'):
+        yield run_unusable_amount_returns_null, arch
+
+
+def run_free_null_preserves_state(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.heap.malloc(30)
+    p = s.heap.malloc(40)
+    s.heap.malloc(50)
+    s.heap.free(p)
+    s2 = s.copy()
+    s2.heap.free(0)
+    nose.tools.assert_true(same_heap_states(s, s2))
+
+def test_free_null_preserves_state():
+    for arch in ('X86', 'AMD64'):
+        yield run_free_null_preserves_state, arch
+
+
+def run_skips_chunks_too_small(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.heap.malloc(30)
+    p = s.heap.malloc(50)
+    s.heap.malloc(40)
+    s.heap.free(p)
+    p2 = s.heap.calloc(20, 5)
+    nose.tools.assert_less(p, p2)
+
+def test_skips_chunks_too_small():
+    for arch in ('X86', 'AMD64'):
+        yield run_skips_chunks_too_small, arch
+
+
+def run_skips_chunks_too_small(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.heap.malloc(30)
+    p = s.heap.malloc(50)
+    s.heap.malloc(40)
+    s.heap.free(p)
+    p2 = s.heap.calloc(20, 5)
+    nose.tools.assert_less(p, p2)
+
+def test_skips_chunks_too_small():
+    for arch in ('X86', 'AMD64'):
+        yield run_skips_chunks_too_small, arch
+
+
+def run_calloc_multiplies(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.heap.malloc(30)
+    sc = s.copy()
+    s.heap.malloc(100)
+    sc.heap.calloc(4, 25)
+    nose.tools.assert_true(same_heap_states(s, sc))
+
+def test_calloc_multiplies():
+    for arch in ('X86', 'AMD64'):
+        yield run_calloc_multiplies, arch
+
+
+def run_calloc_clears(arch):
+    s = SimState(arch=arch, heap=SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))
+    s.memory.store(0xd0000000 + 2 * s.heap._chunk_size_t_size, s.solver.BVV(-1, 100 * 8))
+    sc = s.copy()
+    p1 = s.heap.calloc(6, 5)
+    p2 = sc.heap.malloc(30)
+    v1 = s.memory.load(p1, 30)
+    v2 = sc.memory.load(p2, 30)
+    nose.tools.assert_true(s.solver.is_true(v1 == 0))
+    nose.tools.assert_true(sc.solver.is_true(v2 == -1))
+
+def test_calloc_clears():
+    for arch in ('X86', 'AMD64'):
+        yield run_calloc_clears, arch
```

Comment:
Could you add an `if __name__ == '__main__':` clause that runs all the tests?

Reply:
Addressed [here](https://github.com/angr/angr/pull/1432/commits/4b2eb577bfee89b3f0c56853ea0ecdd5539b5650).